language-plutus-core/src/Language/PlutusCore/Constant/Name.hs

```diff
@@ -25,17 +31,33 @@ module Language.PlutusCore.Constant.Name
     , typedGtByteString
     ) where
 
-import           Language.PlutusCore.Constant.Dynamic.Instances     ()
+import           Language.PlutusCore.Constant.Dynamic.Instances ()
 import           Language.PlutusCore.Constant.Typed
 import           Language.PlutusCore.Core
-import           Language.PlutusCore.Evaluation.Machine.ExBudgeting
 import           Language.PlutusCore.Evaluation.Result
+import           Language.PlutusCore.Universe
 
-import qualified Data.ByteString.Lazy.Char8                         as BSL
+import qualified Data.ByteString.Lazy                           as BSL
 import           Data.Proxy
 
+-- | A class that allows to derive a 'TypeScheme' for a builtin.
+class KnownTypeScheme uni as r where
+    knownTypeScheme :: TypeScheme uni as r
+
+instance KnownType uni r => KnownTypeScheme uni '[] r where
+    knownTypeScheme = TypeSchemeResult Proxy
+
+instance (KnownType uni a, KnownTypeScheme uni as r) => KnownTypeScheme uni (a ': as) r where
+    knownTypeScheme = Proxy `TypeSchemeArrow` knownTypeScheme
+
+-- | Automatically typify a 'BuiltinName'.
```

Comment:
Like, inline things like `typedAddInteger` as in

```haskell
withTypedBuiltinName AddInteger           k = k $ makeTypedBuiltinName @[Integer, Integer] @Integer AddInteger
```

or what do you mean? I'm fine either way, so if people prefer to inline the `typed*` definitions, then be it.

Reply:
I'd go with that yeah. I think the type system would complain either way at some point though, so it's fine with me this way as well.

Reply:
My very slight preference is to keep everything the same, because `makeTypedBuiltinName` works only for non-polymorphic builtins, so I kind of like the current general structure where `makeTypeBuiltinName` is only used as a shorthand that allows not to write a `TypeScheme` explicitly, which is still needed in the general case.

So we have 1:1. @michaelpj you decide.

Reply:
Ah, and `withTypedBuiltinName` is supposed to be gone once we unhardcode the built-in functions, which is another reason to keep the `TypeScheme`s separately.

Reply:
Don't we use `typedAddInteger` anyway? I don't think it makes much difference, so if Roman is planning to get rid of this function anyway then we might as well just leave it.

Reply:
> Don't we use `typedAddInteger` elsewhere?

Yes, we use it in the tests, but we could use the more verbose 

```haskell
withTypedBuiltinName AddInteger $ \typedAddInteger -> ...
```

Ok, I'll keep things the same.